src/chart/helper/createSeriesDataFromArray.ts

```diff
@@ -0,0 +1,208 @@
+                dimInfo.createInvertedIndices = true;
+            }
+        }
+        if (dimInfo.otherDims.itemName != null) {
+            hasNameEncode = true;
+        }
+    });
+    if (!hasNameEncode && firstCategoryDimIndex != null) {
+        dimInfoList[firstCategoryDimIndex].otherDims.itemName = 0;
+    }
+    return firstCategoryDimIndex;
+}
+
+function createListFromArray(
+    sourceOrStore: Source | OptionSourceData | DataStorage,
+    seriesModel: SeriesModel,
+    opt?: {
+        generateCoord?: string
+        useEncodeDefaulter?: boolean | EncodeDefaulter
+        // By default: auto. If `true`, create inverted indices for all ordinal dimension on coordSys.
+        createInvertedIndices?: boolean
+    }
+): SeriesData {
+    opt = opt || {};
+
+    if (!isSourceInstance(sourceOrStore) && !isDataStorage(sourceOrStore)) {
+        sourceOrStore = createSourceFromSeriesDataOption(
+            sourceOrStore as OptionSourceData
+        );
+    }
+
+    const source = isDataStorage(sourceOrStore) ? sourceOrStore.getSource() : sourceOrStore;
+    const coordSysInfo = getCoordSysInfoBySeries(seriesModel);
+    const coordSysDimDefs = getCoordSysDimDefs(seriesModel, coordSysInfo);
+    const useEncodeDefaulter = opt.useEncodeDefaulter;
+
+    // Try to ignore unsed dimensions if sharing a high dimension datastorage
+    // 30 is an experience value.
+    const omitUnusedDimensions = isDataStorage(sourceOrStore) && sourceOrStore.getDimensionCount() > 30;
+    const encodeDefaulter = zrUtil.isFunction(useEncodeDefaulter)
+        ? useEncodeDefaulter
+        : useEncodeDefaulter
+        ? zrUtil.curry(makeSeriesEncodeForAxisCoordSys, coordSysDimDefs, seriesModel)
+        : null;
+
+    const createDimensionOptions = {
+        coordDimensions: coordSysDimDefs,
+        generateCoord: opt.generateCoord,
+        encodeDefine: seriesModel.getEncode()
+            // NOTE: If we call createDimensions on same source multiple times.
+            // It will break the encodeDefaulter which has sideeffects.
+            // So we prepare the default encode here instead of passing encoderDefaulter function.
+            || (encodeDefaulter && encodeDefaulter(
+                source, getDimCount(source, coordSysDimDefs, source.dimensionsDefine || [])
+            )),
+        omitUnusedDimensions
+    };
+    let dimInfoList = createDimensions(sourceOrStore, createDimensionOptions);
+    let firstCategoryDimIndex = injectOrdinalMeta(dimInfoList, opt.createInvertedIndices, coordSysInfo);
+
+    if (omitUnusedDimensions) {
+        // sourceOrStore
+        if (!(sourceOrStore as DataStorage).syncDimensionTypes(dimInfoList)) {
+            // We need a full dimensions list if we want to recreate the storage.
+            // Or the storage will ignore the data of other dimensions.
+            dimInfoList = createDimensions(sourceOrStore, zrUtil.extend(createDimensionOptions, {
+                omitUnusedDimensions: true
```

Comment:
Fixed. 

In fact, after fixing this bug, I noticed stacked high dimension data still has a serious performance issue because the stack dimensions will cause `canUse`(syncDimensionsInfo previously) to fail and DataStorage creation. This makes me have to rethink the design.